src/core/batch.rkt

```diff
@@ -145,6 +153,86 @@
+        [(list '$approx spec impl)
+         (define spec-type (if (representation? type) (representation-type type) type))
+         (approx (loop spec spec-type) (loop impl type))]
+        [(list 'if cond ift iff)
+         (if (representation? type)
+             (list 'if (loop cond (get-representation 'bool)) (loop ift type) (loop iff type))
+             (list 'if (loop cond 'bool) (loop ift type) (loop iff type)))]
+        [(list (? impl-exists? impl) args ...) (cons impl (map loop args (impl-info impl 'itype)))]
+        [(list op args ...) (cons op (map loop args (operator-info op 'itype)))])))
+
+  (define (eggref id)
+    (cdr (vector-ref egg-nodes id)))
+
+  (define (add-root node type)
+    (match node
+      [(? number?)
+       (batch-push! out (if (representation? type) (literal node (representation-name type)) node))]
+      [(? symbol?)
+       (batch-push! out (if (hash-has-key? rename-dict node) (car (hash-ref rename-dict node)) node))]
+      [(list '$approx spec impl)
+       (define spec* (vector-ref id->spec spec))
+       (unless spec*
+         (error 'regraph-extract-variants "no initial approx node in eclass"))
+       (define spec-type (if (representation? type) (representation-type type) type))
+       (define final-spec (egg-parsed->expr spec* rename-dict spec-type))
+       (batch-push! out (approx final-spec (add-root (eggref impl) type)))]
+      [(list 'if cond ift iff)
+       (define cond-node (eggref cond))
+       (define ift-node (eggref ift))
+       (define iff-node (eggref iff))
+       (if (representation? type)
+           (batch-push! out
+                        (list 'if
+                              (add-root cond-node (get-representation 'bool))
+                              (add-root ift-node type)
+                              (add-root iff-node type)))
+           (batch-push!
+            out
+            (list 'if (add-root cond-node 'bool) (add-root ift-node type) (add-root iff-node type))))]
+      [(list (? impl-exists? impl) ids ...)
+       (define args
+         (for/list ([id (in-list ids)]
+                    [type (in-list (impl-info impl 'itype))])
+           (add-root (eggref id) type)))
+       (batch-push! out (cons impl args))]
+      [(list (? operator-exists? op) ids ...)
+       (define args
+         (for/list ([id (in-list ids)]
+                    [type (in-list (operator-info op 'itype))])
+           (add-root (eggref id) type)))
+       (batch-push! out (cons op args))]))
+
+  (define (finalize-batch)
+    (set-batch-nodes! input-batch
+                      (vector-append (batch-nodes input-batch)
+                                     (list->vector (reverse (mutable-batch-nodes out))))))
+
+  ; Cleaning the batch to start over
+  (define (clean-batch)
+    (set! out (make-mutable-batch))
+    (set-mutable-batch-index! out (batch-restore-index input-batch)))
+  (values add-root clean-batch finalize-batch))
```

Comment:
I'm a little confused why we need `clean-batch` here.

Reply:
I'm actually too, when I do `regraph-extract-variants` I assume that I start over extracting to a clean batch that contains only the `nodes` that I passed previously.
But turned out that sometimes `regraph-extract-variants` can be called multiple times for the same `egg-typed-batch-extract` object I guess.
And as a result - some unexpected nodes gets added to my batch.
`clean-batch` simply protects from these cases. You start extracting from egg to a batch - clean the batch - extract to the batch - and finalize your extraction by updating the batch that you were given.
I will spend time to find where this problem comes from

Reply:
Seems not to be the case anymore. Removed `clean-batch`. Have not found reasons to have it around